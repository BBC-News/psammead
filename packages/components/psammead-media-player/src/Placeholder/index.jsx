import React from 'react';
import styled from 'styled-components';
import { string, func, shape, oneOf } from 'prop-types';
import Image from '@bbc/psammead-image';
import PlayButton from '@bbc/psammead-play-button';
import { C_POSTBOX } from '@bbc/psammead-styles/colours';
import Guidance from '../Guidance';

const StyledPlaceholder = styled.div`
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
`;

const StyledPlayButton = styled(PlayButton)`
  position: absolute;
  bottom: 0;

  /* stylelint-disable */
  /* https://www.styled-components.com/docs/advanced#referring-to-other-components */
  ${StyledPlaceholder}:hover &,
  ${StyledPlaceholder}:focus & {
    background-color: ${C_POSTBOX};
  }
  /* stylelint-enable */
`;

const Placeholder = ({ onClick, service, src, srcset, mediaInfo }) => {
  const {
    title,
    datetime,
    duration,
    durationSpoken,
    type,
    guidanceMessage,
  } = mediaInfo;

  return (
    <StyledPlaceholder onClick={onClick}>
      {guidanceMessage && (
        <Guidance service={service} guidanceMessage={guidanceMessage} />
      )}
      <StyledPlayButton
        title={title}
        service={service}
        onClick={() => {}}
        datetime={datetime}
        duration={duration}
        durationSpoken={durationSpoken}
        type={type}
        guidanceMessage={guidanceMessage}
      />
      <Image alt="" src={src} srcset={srcset} />
    </StyledPlaceholder>
  );
};

Placeholder.propTypes = {
  onClick: func.isRequired,
  service: string.isRequired,
  src: string.isRequired,
  srcset: string,
  mediaInfo: shape({
    title: string.isRequired,
    datetime: string,
    duration: string,
    durationSpoken: string,
    type: oneOf(['audio', 'video']),
    guidanceMessage: string,
  }),
};
Placeholder.defaultProps = {
  srcset: null,
  mediaInfo: shape({
    datetime: null,
    duration: null,
    durationSpoken: null,
    type: 'video',
    guidanceMessage: null,
  }),
};

export default Placeholder;
