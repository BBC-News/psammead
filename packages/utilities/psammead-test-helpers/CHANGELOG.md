# Psammead Test Helpers Changelog

| Version | Description |
|---------|-------------|
| 0.3.0   | [PR#224](https://github.com/BBC-News/psammead/pull/224) Adds a helper method for testing package export values and uses it to test its own exports. |
| 0.2.1   | [PR#255](https://github.com/BBC-News/psammead/pull/255) Bumps package version to fix what appears to be an erroneously-published version. |
| 0.1.2   | [PR#212](https://github.com/BBC-News/psammead/pull/212) Update package description and README. |
| 0.1.1   | [PR#173](https://github.com/BBC-News/psammead/pull/173) Update PRs welcome link |
| 0.1.0   | [PR#80](https://github.com/BBC-News/psammead/pull/80) Create initial package, pulled in from [simorgh](https://github.com/BBC-News/simorgh). |
