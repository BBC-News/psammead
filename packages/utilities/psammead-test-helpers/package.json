{
  "name": "@bbc/psammead-test-helpers",
  "version": "0.3.0",
  "main": "dist/index.js",
  "description": "A collection of helper methods for implementing Jest snapshot tests for styled-components, required by many Psammead components.",
  "repository": {
    "type": "git",
    "url": "https://github.com/BBC-News/psammead/tree/latest/packages/utilities/psammead-test-helpers"
  },
  "author": {
    "name": "Psammead Maintainers",
    "email": "PsammeadMaintainers@bbc.co.uk"
  },
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/bbc/psammead/issues"
  },
  "homepage": "https://github.com/BBC-News/psammead/blob/latest/packages/utilities/psammead-test-helpers/README.md",
  "keywords": [
    "bbc",
    "utilities",
    "test",
    "helpers",
    "jest",
    "snapshot"
  ],
  "dependencies": {
    "jest-styled-components": "^6.3.1",
    "react-test-renderer": "^16.6.3"
  }
}
