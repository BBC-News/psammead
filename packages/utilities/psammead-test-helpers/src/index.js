import renderer from 'react-test-renderer';
import 'jest-styled-components';
import ShallowRenderer from 'react-test-renderer/shallow';

export const shouldMatchSnapshot = (title, component) => {
  it(title, () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
};

export const shallowRender = component => {
  const shallowRenderer = new ShallowRenderer();
  shallowRenderer.render(component);
  const result = shallowRenderer.getRenderOutput();

  return result;
};

export const shouldShallowMatchSnapshot = (title, component) => {
  it(title, () => {
    const tree = shallowRender(component);
    expect(tree).toMatchSnapshot();
  });
};

export const isNull = (title, component) => {
  it(title, () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toBeNull();
  });
};

const errorIfMissingKey = (keys, object, message) => {
  keys.forEach(key => {
    if (!(key in object)) {
      throw new Error(`Missing value '${key}' in ${message}.`);
    }
  });
};

const checkKeysExistInBothObjects = (object1, object2, message1, message2) => {
  const object1Keys = Object.keys(object1);
  const object2Keys = Object.keys(object2);

  errorIfMissingKey(object1Keys, object2, message2);
  errorIfMissingKey(object2Keys, object1, message1);
};

const checkTypesOfExports = (
  actualExportsByName,
  actualExports,
  expectedExports,
  utilityName,
) => {
  actualExportsByName.forEach(actualExportName => {
    const actualExportValue = actualExports[utilityName][actualExportName];
    const expectedExport = expectedExports[utilityName][actualExportName];
    const typeCheck = typeof actualExportValue === expectedExport; // eslint-disable-line valid-typeof

    // if this fails it is likely that an export is missing from the unit test expectation
    expect(typeCheck).toBe(true);
  });
};

export const testUtilityPackages = (
  actualExports,
  expectedExports,
  packageName,
) => {
  const actualUtilities = Object.keys(actualExports);

  // check if the actual exported file has defined expected values to match and that the expected values are actually exported
  checkKeysExistInBothObjects(
    actualExports,
    expectedExports,
    `the actual utilities for '${packageName}'`,
    `the expected utilities for '${packageName}'`,
  );

  actualUtilities.forEach(utilityName => {
    const actualExportsByName = Object.keys(actualExports[utilityName]);

    // check if each of the actual exported consts have an expected value to match and that the expected values are actually exported
    checkKeysExistInBothObjects(
      actualExports[utilityName],
      expectedExports[utilityName],
      `the actual export for '${packageName}/${utilityName}'`,
      `the expected export for '${packageName}/${utilityName}'`,
    );

    checkTypesOfExports(
      actualExportsByName,
      actualExports,
      expectedExports,
      utilityName,
    );
  });
};
