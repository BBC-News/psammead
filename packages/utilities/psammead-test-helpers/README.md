# psammead-test-helpers &middot; [![GitHub license](https://img.shields.io/badge/license-Apache%202.0-blue.svg)](https://github.com/BBC-News/psammead/blob/latest/LICENSE) [![npm version](https://img.shields.io/npm/v/@bbc/psammead-test-helpers.svg)](https://www.npmjs.com/package/@bbc/psammead-test-helpers) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://github.com/BBC-News/psammead/blob/latest/CONTRIBUTING.md)

This package is a collection helper methods used for snapshot testing, with support for styled components.

## Installation

```jsx
npm install react react-dom @bbc/psammead-test-helpers --save-dev
```

## Usage

```jsx
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';

shouldMatchSnapshot(
  'should render correctly',
  <h1>Hello World</h1>,
);
```

## Functions

| Name                       | Arguments        | Description  |
|:---------------------------|:-----------------|:-------------|
| shouldMatchSnapshot        | title, component | Renders the component using react-test-renderer, converts it to JSON and asserts that it matches the given snapshot, which will be saved in the `__snapshots__` directory. The first argument `title` is the title for the test. |
| shallowRender              | component        |  Shallow renders the component using react-test-renderer and returns the render output |
| shouldShallowMatchSnapshot | title, component | Shallow renders the component using react-test-renderer and asserts that it matches the given snapshot, which will be saved in the `__snapshots__` directory. The first argument `title` is the title for the test. |
| isNull                     | title, component    | Renders the component using react-test-renderer, converts it to JSON and asserts that it is null. The first argument `title` is the title for the test. |

## Roadmap

There is currently a plan to migrate from `react-test-renderer` to `enzyme` as it provides a nicer higher-level API over the same functionality, as well as having support for more granular unit tests.

## Contributing

Psammead is completely open source. We are grateful for any contributions, whether they be new components, bug fixes or general improvements. Please see our primary contributing guide which can be found at [the root of the Psammead respository](https://github.com/BBC-News/psammead/blob/latest/CONTRIBUTING.md).

### [Code of Conduct](https://github.com/BBC-News/psammead/blob/latest/CODE_OF_CONDUCT.md)

We welcome feedback and help on this work. By participating in this project, you agree to abide by the [code of conduct](https://github.com/BBC-News/psammead/blob/latest/CODE_OF_CONDUCT.md). Please take a moment to read it.

### License

Psammead is [Apache 2.0 licensed](https://github.com/BBC-News/psammead/blob/latest/LICENSE).
