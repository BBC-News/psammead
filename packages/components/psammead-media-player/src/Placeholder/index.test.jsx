import React from 'react';
import { render, fireEvent, getByText } from '@testing-library/react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import Placeholder from '.';

describe('Media Player: Placeholder', () => {
  const mockOnClick = jest.fn();
  const withDuration = {
    duration: '2:30',
    durationSpoken: '2 minutes 30 seconds',
    datetime: 'PT2M30S',
  };

  afterEach(() => {
    jest.resetModules();
  });

  shouldMatchSnapshot(
    'should render a video placeholder',
    <Placeholder
      onClick={mockOnClick}
      src="http://foo.bar/placeholder.png"
      service="news"
      mediaInfo={{ title: 'Dog chases cat.', ...withDuration }}
      className="foo"
    />,
  );

  shouldMatchSnapshot(
    'should render a video placeholder without duration',
    <Placeholder
      onClick={mockOnClick}
      src="http://foo.bar/placeholder.png"
      service="news"
      mediaInfo={{ title: 'Dog chases cat.' }}
      className="foo"
    />,
  );

  shouldMatchSnapshot(
    'should render an audio placeholder',
    <Placeholder
      onClick={mockOnClick}
      src="http://foo.bar/placeholder.png"
      service="news"
      mediaInfo={{ type: 'audio', title: 'Dog barks at cat.', ...withDuration }}
      className="foo"
    />,
  );

  shouldMatchSnapshot(
    'should render an audio placeholder without duration',
    <Placeholder
      onClick={mockOnClick}
      src="http://foo.bar/placeholder.png"
      service="news"
      mediaInfo={{ type: 'audio', title: 'Dog barks at cat.' }}
      className="foo"
    />,
  );

  it('should call onClick when the placeholder and play button is clicked', () => {
    const { container } = render(
      <Placeholder
        onClick={mockOnClick}
        src="http://foo.bar/placeholder.png"
        service="news"
        mediaInfo={{ title: 'Dog chases cat.', ...withDuration }}
        className="foo"
      />,
    );
    fireEvent.click(container.firstChild);
    fireEvent.click(getByText(container.firstChild, '2:30'));
    expect(mockOnClick).toHaveBeenCalledTimes(2);
  });
});
