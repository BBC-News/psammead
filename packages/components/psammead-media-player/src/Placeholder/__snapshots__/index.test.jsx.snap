// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Media Player: Placeholder should render a video placeholder 1`] = `
.c8 {
  display: block;
  width: 100%;
  visibility: visible;
}

.c6 {
  vertical-align: middle;
  margin: 0 0.25rem;
  color: #222222;
  fill: currentColor;
  width: 0.75rem;
  height: 0.75rem;
}

.c4 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

.c2 {
  background-color: #222222;
  border: none;
  color: #FFFFFF;
  cursor: pointer;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 700;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  height: 5rem;
  padding: 0;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
  width: 5rem;
}

.c2:hover,
.c2:focus {
  background-color: #B80000;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
}

.c5 > svg {
  color: #FFFFFF;
  fill: currentColor;
  height: 1.5rem;
  margin-top: 0.5rem;
  width: 1.5rem;
}

.c7 {
  display: block;
  margin-top: 0.5rem;
}

.c1 {
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c3 {
  position: absolute;
  bottom: 0;
}

.c0:hover .c3,
.c0:focus .c3 {
  background-color: #B80000;
}

<div
  class="c0 c1"
>
  <button
    class="c2 c3"
  >
    <span
      class="c4"
    >
      Play video, "Dog chases cat.", 2 minutes 30 seconds
    </span>
    <div
      aria-hidden="true"
      class="c5"
    >
      <svg
        class="c6"
        focusable="false"
        height="12px"
        viewBox="0 0 32 32"
        width="12px"
      >
        <polygon
          points="3,32 29,16 3,0"
        />
      </svg>
    </div>
    <time
      aria-hidden="true"
      class="c7"
      datetime="PT2M30S"
    >
      2:30
    </time>
  </button>
  <img
    alt=""
    class="c8"
    src="http://foo.bar/placeholder.png"
  />
</div>
`;

exports[`Media Player: Placeholder should render a video placeholder with guidance 1`] = `
.c10 {
  display: block;
  width: 100%;
  visibility: visible;
}

.c8 {
  vertical-align: middle;
  margin: 0 0.25rem;
  color: #222222;
  fill: currentColor;
  width: 0.75rem;
  height: 0.75rem;
}

.c6 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

.c4 {
  background-color: #222222;
  border: none;
  color: #FFFFFF;
  cursor: pointer;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 700;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  height: 5rem;
  padding: 0;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
  width: 5rem;
}

.c4:hover,
.c4:focus {
  background-color: #B80000;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
}

.c7 > svg {
  color: #FFFFFF;
  fill: currentColor;
  height: 1.5rem;
  margin-top: 0.5rem;
  width: 1.5rem;
}

.c9 {
  display: block;
  margin-top: 0.5rem;
}

.c2 {
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 400;
  font-style: normal;
  font-size: 0.9375rem;
  line-height: 1.125rem;
  width: 100%;
  height: 100%;
  position: absolute;
  background-color: rgba(34,34,34,0.75);
  border: 0.0625rem solid transparent;
  color: #FFFFFF;
}

.c3 {
  display: block;
  font-weight: normal;
  padding: 0.5rem;
  border-bottom: 0.0625rem solid transparent;
}

.c1 {
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c5 {
  position: absolute;
  bottom: 0;
}

.c0:hover .c5,
.c0:focus .c5 {
  background-color: #B80000;
}

@media (min-width:37.5rem) {
  .c2 {
    font-size: 0.875rem;
  }
}

@media screen and (-ms-high-contrast:active) {
  .c2 {
    background-color: transparent;
  }
}

@media screen and (-ms-high-contrast:active) {
  .c3 {
    background-color: window;
  }
}

@media (min-width:25rem) {
  .c3 {
    padding: 1rem;
  }
}

<div
  class="c0 c1"
>
  <div
    class="c2"
  >
    <strong
      aria-hidden="true"
      class="c3"
    >
      Guidance: May contain strong language, sexual or violent content that may offend.
    </strong>
  </div>
  <button
    class="c4 c5"
  >
    <span
      class="c6"
    >
      Guidance: May contain strong language, sexual or violent content that may offend. Play video, "Dog chases cat.", 2 minutes 30 seconds
    </span>
    <div
      aria-hidden="true"
      class="c7"
    >
      <svg
        class="c8"
        focusable="false"
        height="12px"
        viewBox="0 0 32 32"
        width="12px"
      >
        <polygon
          points="3,32 29,16 3,0"
        />
      </svg>
    </div>
    <time
      aria-hidden="true"
      class="c9"
      datetime="PT2M30S"
    >
      2:30
    </time>
  </button>
  <img
    alt=""
    class="c10"
    src="http://foo.bar/placeholder.png"
  />
</div>
`;

exports[`Media Player: Placeholder should render a video placeholder without duration 1`] = `
.c7 {
  display: block;
  width: 100%;
  visibility: visible;
}

.c6 {
  vertical-align: middle;
  margin: 0 0.25rem;
  color: #222222;
  fill: currentColor;
  width: 0.75rem;
  height: 0.75rem;
}

.c4 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

.c2 {
  background-color: #222222;
  border: none;
  color: #FFFFFF;
  cursor: pointer;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 700;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  height: 5rem;
  padding: 0;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
  width: 5rem;
}

.c2:hover,
.c2:focus {
  background-color: #B80000;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
}

.c5 > svg {
  color: #FFFFFF;
  fill: currentColor;
  height: 1.5rem;
  width: 1.5rem;
}

.c1 {
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c3 {
  position: absolute;
  bottom: 0;
}

.c0:hover .c3,
.c0:focus .c3 {
  background-color: #B80000;
}

<div
  class="c0 c1"
>
  <button
    class="c2 c3"
  >
    <span
      class="c4"
    >
      Play video, "Dog chases cat."
    </span>
    <div
      aria-hidden="true"
      class="c5"
    >
      <svg
        class="c6"
        focusable="false"
        height="12px"
        viewBox="0 0 32 32"
        width="12px"
      >
        <polygon
          points="3,32 29,16 3,0"
        />
      </svg>
    </div>
  </button>
  <img
    alt=""
    class="c7"
    src="http://foo.bar/placeholder.png"
  />
</div>
`;

exports[`Media Player: Placeholder should render an audio placeholder 1`] = `
.c8 {
  display: block;
  width: 100%;
  visibility: visible;
}

.c6 {
  vertical-align: middle;
  margin: 0 0.25rem;
  color: #222222;
  fill: currentColor;
  width: 0.8125rem;
  height: 0.75rem;
}

.c4 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

.c2 {
  background-color: #222222;
  border: none;
  color: #FFFFFF;
  cursor: pointer;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 700;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  height: 5rem;
  padding: 0;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
  width: 5rem;
}

.c2:hover,
.c2:focus {
  background-color: #B80000;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
}

.c5 > svg {
  color: #FFFFFF;
  fill: currentColor;
  height: 1.5rem;
  margin-top: 0.5rem;
  width: 1.5rem;
}

.c7 {
  display: block;
  margin-top: 0.5rem;
}

.c1 {
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c3 {
  position: absolute;
  bottom: 0;
}

.c0:hover .c3,
.c0:focus .c3 {
  background-color: #B80000;
}

<div
  class="c0 c1"
>
  <button
    class="c2 c3"
  >
    <span
      class="c4"
    >
      Play audio, "Dog barks at cat.", 2 minutes 30 seconds
    </span>
    <div
      aria-hidden="true"
      class="c5"
    >
      <svg
        class="c6"
        focusable="false"
        height="12px"
        viewBox="0 0 13 12"
        width="13px"
      >
        <path
          d="M9.021 1.811l-.525.525c.938.938 1.5 2.25 1.5 3.675s-.563 2.738-1.5 3.675l.525.525c1.05-1.087 1.725-2.55 1.725-4.2s-.675-3.112-1.725-4.2z"
        />
        <path
          d="M10.596.199l-.525.562c1.35 1.35 2.175 3.225 2.175 5.25s-.825 3.9-2.175 5.25l.525.525c1.5-1.462 2.4-3.525 2.4-5.775s-.9-4.312-2.4-5.812zM6.996 1.511l-2.25 2.25H.996v4.5h3.75l2.25 2.25z"
        />
      </svg>
    </div>
    <time
      aria-hidden="true"
      class="c7"
      datetime="PT2M30S"
    >
      2:30
    </time>
  </button>
  <img
    alt=""
    class="c8"
    src="http://foo.bar/placeholder.png"
  />
</div>
`;

exports[`Media Player: Placeholder should render an audio placeholder without duration 1`] = `
.c7 {
  display: block;
  width: 100%;
  visibility: visible;
}

.c6 {
  vertical-align: middle;
  margin: 0 0.25rem;
  color: #222222;
  fill: currentColor;
  width: 0.8125rem;
  height: 0.75rem;
}

.c4 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
  margin: 0;
}

.c2 {
  background-color: #222222;
  border: none;
  color: #FFFFFF;
  cursor: pointer;
  display: block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-weight: 700;
  font-style: normal;
  font-size: 0.75rem;
  line-height: 1rem;
  height: 5rem;
  padding: 0;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
  width: 5rem;
}

.c2:hover,
.c2:focus {
  background-color: #B80000;
  -webkit-transition: background-color 300ms;
  transition: background-color 300ms;
}

.c5 > svg {
  color: #FFFFFF;
  fill: currentColor;
  height: 1.5rem;
  width: 1.5rem;
}

.c1 {
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c3 {
  position: absolute;
  bottom: 0;
}

.c0:hover .c3,
.c0:focus .c3 {
  background-color: #B80000;
}

<div
  class="c0 c1"
>
  <button
    class="c2 c3"
  >
    <span
      class="c4"
    >
      Play audio, "Dog barks at cat."
    </span>
    <div
      aria-hidden="true"
      class="c5"
    >
      <svg
        class="c6"
        focusable="false"
        height="12px"
        viewBox="0 0 13 12"
        width="13px"
      >
        <path
          d="M9.021 1.811l-.525.525c.938.938 1.5 2.25 1.5 3.675s-.563 2.738-1.5 3.675l.525.525c1.05-1.087 1.725-2.55 1.725-4.2s-.675-3.112-1.725-4.2z"
        />
        <path
          d="M10.596.199l-.525.562c1.35 1.35 2.175 3.225 2.175 5.25s-.825 3.9-2.175 5.25l.525.525c1.5-1.462 2.4-3.525 2.4-5.775s-.9-4.312-2.4-5.812zM6.996 1.511l-2.25 2.25H.996v4.5h3.75l2.25 2.25z"
        />
      </svg>
    </div>
  </button>
  <img
    alt=""
    class="c7"
    src="http://foo.bar/placeholder.png"
  />
</div>
`;
